feat: allow specifying custom relay in dumbpipe (#60)

alx0x10 opened the pull request:
Hello I'm back!

I thought this issue would be a good first step to get familiar with Iroh, so I started implementing the possibility to define a custom relay.  
As mentioned in the issue, Sendme already supports this, so I mostly adapted that code.

Before going further and adding tests, I’d like your feedback on a few points:

- Do you think this feature makes sense here? Since Sendme already demonstrates it, maybe you prefer to keep dumbpipe shorter and focused only on the listen/connect logic?  
- I introduced the `anyhow` dependency like in Sendme. Would you prefer a different error-handling approach?
- Like in Sendme, I wrapped `RelayMode` into `RelayModeOption`, which removes the `Staging` option. I haven’t really looked into how staging works, would you like me to explore that for this implementation?

eminence commented on Cargo.toml:
```diff
@@ -27,6 +27,7 @@ tracing = "0.1.40"
 tracing-subscriber = { version = "0.3.18", features = ["env-filter"] }
 data-encoding = "2.9.0"
 n0-snafu = "0.2.1"
+anyhow = "1.0.99"
```
It doesn't look like `anyhow` is used in your PR?

alx0x10 commented:
Ok, thank you for the feedback, I hadn't seen this RelayUrlParseError.
I made the modification and I also prefer it much more like that!

I'll look into the testing soon then and come back to you :+1:

dignifiedquire commented:
This looks pretty good, can you fix the lint issues and move it to ready (it is still marked as draft)?

alx0x10 commented:
Actually, as @eminence says, when running dumbpipe with the `--relay disable`, the code blocks at `endpoint.home_relay().initialized().await;` since no home relay is initialized.

I added a check to not run this check if we disable the use of a relay. 

```
    if args.common.relay != RelayModeOption::Disabled {
        endpoint.home_relay().initialized().await;
    }
```

And I noticed that this check is not made for the `connect_stdio()`... I couldn't understand the reason for it.

Beside that I added some tests and info in the readme.

eminence commented on src/main.rs:
```diff
@@ -122,6 +123,14 @@ pub struct CommonArgs {
     #[clap(long)]
     pub custom_alpn: Option<String>,
 
+    /// The relay URL to use as a home relay,
+    ///
+    /// Can be set to "disabled" to disable relay servers and "custom"
+    /// to configure custom servers. The default is the n0 quickest responding
+    /// relay if the flag is not set.
+    #[clap(long, default_value_t = RelayModeOption::Default)]
+    pub relay: RelayModeOption,
```
This help text needs a little tweaking I think.

You don't set this option to the literal value of "custom", but rather you give your relay url

alx0x10 commented on src/main.rs:
```diff
@@ -291,14 +343,24 @@ async fn create_endpoint(
+        if !(common.relay == RelayModeOption::Disabled) {
+            endpoint.online().await;
+        }
+    }).await;
```
Btw I'm not sure to understand why just `let _ =` the Result of the timeout instead of `.expect()` or something to stop if the endpoint does get online.
